mobile/mysterium/proposals_manager.go

```diff
@@ -108,52 +109,67 @@ func newProposalsManager(
 	repository proposal.Repository,
 	mysteriumAPI mysteriumAPI,
 	filterPresetStorage *proposal.FilterPresetStorage,
+	cacheTTL time.Duration,
 ) *proposalsManager {
 	return &proposalsManager{
 		repository:          repository,
 		mysteriumAPI:        mysteriumAPI,
 		filterPresetStorage: filterPresetStorage,
+		cacheTTL:            cacheTTL,
 	}
 }
 
 type proposalsManager struct {
 	repository          proposal.Repository
 	cache               []market.ServiceProposal
+	cachedAt            time.Time
+	cacheTTL            time.Duration
 	mysteriumAPI        mysteriumAPI
 	filterPresetStorage *proposal.FilterPresetStorage
 }
 
+func (m *proposalsManager) isCacheStale() bool {
+	return time.Now().After(m.cachedAt.Add(m.cacheTTL))
+}
+
 func (m *proposalsManager) getProposals(req *GetProposalsRequest) (*getProposalsResponse, error) {
 	// Get proposals from cache if exists.
-	if !req.Refresh {
-		cachedProposals := m.getFromCache()
-		if len(cachedProposals) > 0 {
-			return m.mapToProposalsResponse(cachedProposals)
+	if req.Refresh || m.isCacheStale() {
+		apiProposals, err := m.getFromRepository(req.toFilter())
+		if err != nil {
+			return nil, err
 		}
+		m.addToCache(apiProposals)
```

Comment:
In reality this is just a API call throttling. In the spirit of previous implementation I leave it multi-thread non-safe :)